HousePrice Regression

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
#train, test data load
train = pd.read_csv("data/train.csv")
test = pd.read_csv('data/test.csv')

print(train.shape)
print(test.shape)

train.head()

(1460, 81)
(1459, 80)


Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [3]:
test.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0  1461          20       RH         80.0    11622   Pave   NaN      Reg   
1  1462          20       RL         81.0    14267   Pave   NaN      IR1   
2  1463          60       RL         74.0    13830   Pave   NaN      IR1   
3  1464          60       RL         78.0     9978   Pave   NaN      IR1   
4  1465         120       RL         43.0     5005   Pave   NaN      IR1   

  LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence MiscFeature  \
0         Lvl    AllPub  ...         120        0    NaN  MnPrv         NaN   
1         Lvl    AllPub  ...           0        0    NaN    NaN        Gar2   
2         Lvl    AllPub  ...           0        0    NaN  MnPrv         NaN   
3         Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
4         HLS    AllPub  ...         144        0    NaN    NaN         NaN   

  MiscVal MoSold  YrSold  SaleType  SaleCondition  
0       0      6    2010        WD         Normal  
1   12500      6    2010        WD         Normal  
2       0      3    2010        WD         Normal  
3       0      6    2010        WD         Normal  
4       0      1    2010        WD         Normal  

[5 rows x 80 columns]

In [5]:
train.isnull().sum()

Id                 0
MSSubClass         0
MSZoning           0
LotFrontage      259
LotArea            0
                ... 
MoSold             0
YrSold             0
SaleType           0
SaleCondition      0
SalePrice          0
Length: 81, dtype: int64

In [6]:
test.isnull().sum()

Id                 0
MSSubClass         0
MSZoning           4
LotFrontage      227
LotArea            0
                ... 
MiscVal            0
MoSold             0
YrSold             0
SaleType           1
SaleCondition      0
Length: 80, dtype: int64

In [7]:
# 각 column과 SalesPrice 간의 상관관계 계산
# 숫자형 data들만 계산
correlations = train.corr(numeric_only=True)['SalePrice'].sort_values(ascending=False)
print("Correlation between columns and SalePrice:")
print(correlations)

Correlation between columns and SalePrice:
SalePrice        1.000000
OverallQual      0.790982
GrLivArea        0.708624
GarageCars       0.640409
GarageArea       0.623431
TotalBsmtSF      0.613581
1stFlrSF         0.605852
FullBath         0.560664
TotRmsAbvGrd     0.533723
YearBuilt        0.522897
YearRemodAdd     0.507101
GarageYrBlt      0.486362
MasVnrArea       0.477493
Fireplaces       0.466929
BsmtFinSF1       0.386420
LotFrontage      0.351799
WoodDeckSF       0.324413
2ndFlrSF         0.319334
OpenPorchSF      0.315856
HalfBath         0.284108
LotArea          0.263843
BsmtFullBath     0.227122
BsmtUnfSF        0.214479
BedroomAbvGr     0.168213
ScreenPorch      0.111447
PoolArea         0.092404
MoSold           0.046432
3SsnPorch        0.044584
BsmtFinSF2      -0.011378
BsmtHalfBath    -0.016844
MiscVal         -0.021190
Id              -0.021917
LowQualFinSF    -0.025606
YrSold          -0.028923
OverallCond     -0.077856
MSSubClass      -0.084284
EnclosedPorch   -0.12

OverallQual (0.791): 재료 및 마감 품질 1(매우 낮음)부터 10(매우 좋음)까지의 점수
GrLivArea (0.709): 지상층 주거 면적
GarageCars (0.640): 수용 가능 차량 개수
GarageArea (0.623): 차고 면적
TotalBsmtSF (0.614): 지하실 면적
1stFlrSF (0.606): 1층 면적
FullBath (0.561): 욕실의 개수
TotRmsAbvGrd (0.534): 총 방의 개수
YearBuilt (0.523): 주택이 건축된 연도
YearRemodAdd (0.507): 리모델링이 완료된 연도
GarageYrBlt (0.486): 차고가 지어진 연도
MasVnrArea (0.477): 외벽용 벽돌의 면적
Fireplaces (0.467): 벽난로의 개수
BsmtFinSF1 (0.386): 지하실 중 거주 가능한 부분의 면적입니다.

LotFrontage (0.352): 주택이 도로와 접하는 면적
WoodDeckSF (0.324): 나무 데크의 면적
2ndFlrSF (0.319): 2층 면적
OpenPorchSF (0.316): 지붕이 있는 현관 면적
HalfBath (0.284): 절반짜리 욕실 (세면대와 변기만 있는)의 개수
LotArea (0.264): 부지(땅)의 면적
BedroomAbvGr (0.168): 지상층 침실의 개수
ScreenPorch (0.111): 스크린이 있는 현관 면적
PoolArea (0.092): 수영장의 면적
MoSold (0.046): 주택이 판매된 월
3SsnPorch (0.045): 3계절용 현관 면적
BsmtFinSF2 (-0.011): 지하실 중 2차 거주 가능한 부분의 면적
BsmtHalfBath (-0.017): 지하실의 절반짜리 욕실 개수
MiscVal (-0.021): 기타 잡다한 시설의 가치
Id (-0.022): 주택을 식별하는 고유 번호입
LowQualFinSF (-0.026): 품질이 낮은 마감된 면적
YrSold (-0.029): 주택이 판매된 연도
OverallCond (-0.078): 주택의 전반적인 상태
MSSubClass (-0.084): 주택의 유형
EnclosedPorch (-0.129): 지붕과 벽이 있는 현관 면적
KitchenAbvGr (-0.136): 부엌의 개수

숫자형 column중 0.4 미만 상관관계 column 삭제

In [8]:
# 상관관계 절댓값이 0.4 미만인 컬럼들을 삭제
abs_correlation_threshold = 0.4
weak_correlated_features = correlations[abs(correlations) < abs_correlation_threshold].index

# 해당 컬럼들을 데이터에서 삭제
# inplace=True를 사용해 원본 데이터프레임을 바로 수정합니다.
train.drop(weak_correlated_features, axis=1, inplace=True)

# 남은 컬럼 목록 확인
print("삭제 후 남은 컬럼 목록:")
print(train.columns)

삭제 후 남은 컬럼 목록:
Index(['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities',
       'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2',
       'BldgType', 'HouseStyle', 'OverallQual', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2',
       'TotalBsmtSF', 'Heating', 'HeatingQC', 'CentralAir', 'Electrical',
       '1stFlrSF', 'GrLivArea', 'FullBath', 'KitchenQual', 'TotRmsAbvGrd',
       'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType', 'GarageYrBlt',
       'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond',
       'PavedDrive', 'PoolQC', 'Fence', 'MiscFeature', 'SaleType',
       'SaleCondition', 'SalePrice'],
      dtype='object')


In [9]:
# 상관관계 절댓값이 0.4 미만인 컬럼들을 삭제
abs_correlation_threshold = 0.4
weak_correlated_features = correlations[abs(correlations) < abs_correlation_threshold].index

# 해당 컬럼들을 데이터에서 삭제
# inplace=True를 사용해 원본 데이터프레임을 바로 수정합니다.
test.drop(weak_correlated_features, axis=1, inplace=True)

# 남은 컬럼 목록 확인
print("삭제 후 남은 컬럼 목록:")
print(test.columns)

삭제 후 남은 컬럼 목록:
Index(['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities',
       'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2',
       'BldgType', 'HouseStyle', 'OverallQual', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2',
       'TotalBsmtSF', 'Heating', 'HeatingQC', 'CentralAir', 'Electrical',
       '1stFlrSF', 'GrLivArea', 'FullBath', 'KitchenQual', 'TotRmsAbvGrd',
       'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType', 'GarageYrBlt',
       'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond',
       'PavedDrive', 'PoolQC', 'Fence', 'MiscFeature', 'SaleType',
       'SaleCondition'],
      dtype='object')


In [11]:
# 숫자형 컬럼만 선택
numeric_train = train.select_dtypes(include=['number'])

# 숫자형 컬럼의 결측치 개수 확인 및 출력
print(numeric_train.isnull().sum())

OverallQual      0
YearBuilt        0
YearRemodAdd     0
MasVnrArea       8
TotalBsmtSF      0
1stFlrSF         0
GrLivArea        0
FullBath         0
TotRmsAbvGrd     0
Fireplaces       0
GarageYrBlt     81
GarageCars       0
GarageArea       0
SalePrice        0
dtype: int64


비슷한 특성끼리 묶어 새로운 feature 만들기

In [14]:
#차고, 주거 면적, 1층 면적 합쳐서 사용
# 방 개수 욕실 개수 합쳐서
# 완공 연도, 리모델링 연도, 지어진 연도 활용해 새로운 feature
# 결측값을 처리합니다.

# 총 면적: 지상층 면적 + 지하실 면적
train['TotalArea'] = train['GrLivArea'] + train['TotalBsmtSF']

# 총 방 개수: 지상층의 총 방 + 욕실의 개수
train['TotalRoomsAndBaths'] = train['TotRmsAbvGrd'] + train['FullBath']

# 주택 나이: 판매 연도 - 건축 연도
# train['HouseAge'] = train['YrSold'] - train['YearBuilt']
#
# # 리모델링 후 경과 연도: 판매 연도 - 리모델링 연도
# train['YearsSinceRemod'] = train['YrSold'] - train['YearRemodAdd']

# 새로 만든 특징들과 'SalePrice' 간의 상관관계 계산
new_features = ['TotalArea', 'TotalRoomsAndBaths']
correlations_with_sale_price = train.corr(numeric_only=True)['SalePrice'][new_features].sort_values(ascending=False)

# 결과를 보기 좋게 출력합니다.
print("새로 만든 특징들과 SalePrice의 상관관계:")
print(correlations_with_sale_price)

새로 만든 특징들과 SalePrice의 상관관계:
TotalArea             0.778959
TotalRoomsAndBaths    0.592733
Name: SalePrice, dtype: float64


In [9]:
numeric_ = X.select_dtypes(exclude=['object']).drop(['MSSubClass'], axis=1).copy()
numeric_.columns

Index(['Id', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond',
       'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2',
       'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF',
       'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath',
       'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces',
       'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF',
       'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'MiscVal',
       'MoSold', 'YrSold'],
      dtype='object')

In [10]:
disc_num_var = ['OverallQual','OverallCond','BsmtFullBath','BsmtHalfBath','FullBath','HalfBath',
                'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces', 'GarageCars', 'MoSold', 'YrSold']

cont_num_var = []
for i in numeric_.columns:
    if i not in disc_num_var:
        cont_num_var.append(i)

In [11]:
cat_train = X.select_dtypes(include=['object']).copy()
cat_train['MSSubClass'] = X['MSSubClass']   #MSSubClass is nominal
cat_train.columns

Index(['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities',
       'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2',
       'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st',
       'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation',
       'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2',
       'Heating', 'HeatingQC', 'CentralAir', 'Electrical', 'KitchenQual',
       'Functional', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual',
       'GarageCond', 'PavedDrive', 'PoolQC', 'Fence', 'MiscFeature',
       'SaleType', 'SaleCondition', 'MSSubClass'],
      dtype='object')

In [ ]:
fig = plt.figure(figsize=(18,16))
for index,col in enumerate(cont_num_var):
    plt.subplot(6,4,index+1)
    sns.distplot(numeric_.loc[:,col].dropna(), kde=False)
fig.tight_layout(pad=1.0)
plt.show()
plt.close()